### ISSUE TITLE
error: ‘constinit’ variable ‘_ProtobufCFileOptions_default_instance_’ does not have a constant initializer

### ISSUE THREAD

**Issue by marxin** [protocolbuffers/protobuf#9916]
Happens with `GCC 12.0.1` due to the use of `__constinit` here:
https://github.com/protocolbuffers/protobuf/blob/67823fe8dc7d57ca33689817c233dcf4f177fa4b/src/google/protobuf/port_def.inc#L653-L655

```
g++ -std=c++11 -DHAVE_CONFIG_H -I.  -include ./config.h -I./protobuf-c -I. -I.    -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -Werror=return-type  -g -MT protobuf-c/protoc_c_protoc_gen_c-protobuf-c.pb.o -MD -MP -MF protobuf-c/.deps/protoc_c_protoc_gen_c-protobuf-c.pb.Tpo -c -o protobuf-c/protoc_c_protoc_gen_c-protobuf-c.pb.o `test -f 'protobuf-c/protobuf-c.pb.cc' || echo './'`protobuf-c/protobuf-c.pb.cc
protobuf-c/protobuf-c.pb.cc:39:124: error: ‘constinit’ variable ‘_ProtobufCFileOptions_default_instance_’ does not have a constant initializer
   39 | PROTOBUF_ATTRIBUTE_NO_DESTROY PROTOBUF_CONSTINIT PROTOBUF_ATTRIBUTE_INIT_PRIORITY1 ProtobufCFileOptionsDefaultTypeInternal _ProtobufCFileOptions_default_instance_;
      |                                                                                                                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
protobuf-c/protobuf-c.pb.cc:39:124: error: ‘ProtobufCFileOptionsDefaultTypeInternal{ProtobufCFileOptionsDefaultTypeInternal::<unnamed union>{ProtobufCFileOptions{google::protobuf::Message{google::protobuf::MessageLite{((& ProtobufCFileOptions::_ZTV20ProtobufCFileOptions) + 16), google::protobuf::internal::InternalMetadata{0}}}, google::protobuf::internal::HasBits<1>{uint32_t [1]()}, google::protobuf::internal::CachedSize{std::atomic<int>{std::__atomic_base<int>{0}}}, google::protobuf::internal::ArenaStringPtr{google::protobuf::internal::TaggedStringPtr{((void*)(& google::protobuf::internal::fixed_address_empty_string))}}, false, false, false, true, true}}}’ is not a constant expression
protobuf-c/protobuf-c.pb.cc:53:127: error: ‘constinit’ variable ‘_ProtobufCMessageOptions_default_instance_’ does not have a constant initializer
   53 | PROTOBUF_ATTRIBUTE_NO_DESTROY PROTOBUF_CONSTINIT PROTOBUF_ATTRIBUTE_INIT_PRIORITY1 ProtobufCMessageOptionsDefaultTypeInternal _ProtobufCMessageOptions_default_instance_;
      |                                                                                                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
protobuf-c/protobuf-c.pb.cc:53:127: error: ‘ProtobufCMessageOptionsDefaultTypeInternal{ProtobufCMessageOptionsDefaultTypeInternal::<unnamed union>{ProtobufCMessageOptions{google::protobuf::Message{google::protobuf::MessageLite{((& ProtobufCMessageOptions::_ZTV23ProtobufCMessageOptions) + 16), google::protobuf::internal::InternalMetadata{0}}}, google::protobuf::internal::HasBits<1>{uint32_t [1]()}, google::protobuf::internal::CachedSize{std::atomic<int>{std::__atomic_base<int>{0}}}, google::protobuf::internal::ArenaStringPtr{google::protobuf::internal::TaggedStringPtr{0}}, false, true}}}’ is not a constant expression
protobuf-c/protobuf-c.pb.cc:65:125: error: ‘constinit’ variable ‘_ProtobufCFieldOptions_default_instance_’ does not have a constant initializer
   65 | PROTOBUF_ATTRIBUTE_NO_DESTROY PROTOBUF_CONSTINIT PROTOBUF_ATTRIBUTE_INIT_PRIORITY1 ProtobufCFieldOptionsDefaultTypeInternal _ProtobufCFieldOptions_default_instance_;
      |                                                                                                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
protobuf-c/protobuf-c.pb.cc:65:125: error: ‘ProtobufCFieldOptionsDefaultTypeInternal{ProtobufCFieldOptionsDefaultTypeInternal::<unnamed union>{ProtobufCFieldOptions{google::protobuf::Message{google::protobuf::MessageLite{((& ProtobufCFieldOptions::_ZTV21ProtobufCFieldOptions) + 16), google::protobuf::internal::InternalMetadata{0}}}, google::protobuf::internal::HasBits<1>{uint32_t [1]()}, google::protobuf::internal::CachedSize{std::atomic<int>{std::__atomic_base<int>{0}}}, false}}}’ is not a constant expression
make: *** [Makefile:1769: protobuf-c/protoc_c_protoc_gen_c-protobuf-c.pb.o] Error 1
```

Note it works if I use `-std=c++17`, so I'm curious if it was ever tested with GCC 12+?

**Comment by marxin** [comment]
For being sure, I've just created an upstream issue for that:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=105491

**Comment by acozzette** [comment]
I'm not sure what the right solution is but I would be happy to review a pull request if anyone has any ideas.

**Comment by yanxue-intel** [comment]
I also met the same build error when using GCC 12.

**Comment by marxin** [comment]
Btw. can you please CC the author of the change, I see the `__constinit` was added in ab4585a6956675ce14a1cba5d321fde980bbf12b, but it's only a Sync from Piper, which is likely a downstream repository?

**Comment by acozzette** [comment]
I believe @dlj-NaN added this and may know more about it.

**Comment by marxin** [comment]
It seems it's a GCC bug that will be fixed in `12.2`.

**Comment by acozzette** [comment]
@marxin Thanks! I tweaked the logic so that we require GCC 12.2 or higher to use `__constinit`, so this should solve the problem.
